client/src/socket.js

```diff
@@ -136,84 +138,77 @@ class HorizonSocket extends WebSocketSubject {
-    // since we don't care about responses. All that matters is that
-    // we can stop it later when needed. If an error to a keepalive
-    // comes back, that's fine since either the socket is down (which
-    // we'll find out about elsewhere) or the keepalive "failed" which
-    // is a server bug, but indicates the socket is still alive, so
-    // it's not really a failure.
+    // Start the keepalive. While it has a request_id, we don't
+    // actually care about responses.
     const keepAliveSub = Observable
             .timer(this.keepalive * 1000, this.keepalive * 1000)
-            .map(n => ({ type: 'keepalive', n }))
-            .subscribe()
+            .map(n => ({
+              type: 'keepalive',
+              n,
+              request_id: this.requestCounter++,
+            })).subscribe(ka => this.next(ka))
     // Make sure keepalive is killed when the handshake is cleaned up
     this.handshakeSub.add(keepAliveSub)
   }
 
+  // Wrapper around the superclass's version of multiplex. With the
+  // following additional features we need for horizon's protocol:
+  // * Generate a request id and filter by it
+  // * Sends handshake on subscription if it hasn't happened already
+  // * Wait for the handshake to complete before sending the request
+  // * Errors when a document with an `error` field is received
+  // * Completes when `state: complete` is received
+  // * Emits `state: synced` as a separate document for easy filtering
   multiplex(rawRequest) {
     return new Observable(subscriber => {
       const requestId = this.requestCounter++
       const request = Object.assign({ request_id: requestId }, rawRequest)
 
-      this.activateRequest(requestId, request)
+      const subMsg = () => {
+        this.activateRequest(requestId, request)
+        return request
+      }
+
+      const unsubMsg = () => {
+        this.deactivateRequest(requestId)
+        return { request_id: requestId, type: 'end_subscription' }
+      }
+
       // Ensure the handshake is complete successfully before sending
       // our request. this.handshake is an AsyncSubject which acts
       // like a Promise and caches its value.
       this.handshake.subscribe({
-        // we don't care about the handshake value so no `next`
-        error: err => subscriber.error(err), // TODO: when
-                                             // reconnection
-                                             // implemented, this
-                                             // shouldn't error out.
+        error: err => subscriber.error(err),
         complete: () => this.next(request), // send request on
-                                            // handshake completion
+        // handshake completion
       })
 
-      // Now we subscribe to messages from the server with our
-      // requestId.
-      const subscription = this.subscribe({
+      this.handshake.ignoreElements().concat(
```

Comment:
Ended up needing to emit an extra event on `complete`, since takeWhile doesn't return the matching event